src/host/renderData.cpp

```diff
@@ -87,14 +87,16 @@ std::vector<Viewport> RenderData::GetSelectionRects() noexcept
 //      they're done with any querying they need to do.
 void RenderData::LockConsole() noexcept
 {
-    ::LockConsole();
+    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    gci.LockConsole();
 }
 
 // Method Description:
 // - Unlocks the console after a call to RenderData::LockConsole.
 void RenderData::UnlockConsole() noexcept
 {
-    ::UnlockConsole();

```

Comment:
Yes, this isn't the right place to emit the CTRL event IMO. There are only 3 situations where calling the global `UnlockConsole` makes sense: The console IO thread (and technically only when handling `GenerateConsoleCtrlEvent`), when handling manual input (because of Ctrl+C) or during shutdown.

As far as I can see, `::UnlockConsole` and it calling `ProcessCtrlEvents()` is just a hold-over from v0.5 code which didn't quite get perfectly integrated into v1 and then into v2, because originally only the global `UnlockConsole` function existed and the one on `CONSOLE_INFORMATION` is a much newer C++ addition. Putting it into `::UnlockConsole` initially makes sense IMO, since the check is cheap and it probably reduced binary size.

But if you think about it, it's not frightening at all and it's also not something we need to keep. That's because we _know_ exactly when a Ctrl event was added: Only and exclusively through `HandleCtrlEvent`. That function is only called from 3 places: The `GenerateConsoleCtrlEvent` console API handler, the conhost message handler `HandleKeyEvent` or in response to closing the conhost window via `CloseConsoleProcessState`.

(Side note: I'm now confused why ConPTY doesn't call `CloseConsoleProcessState`. That's probably not good...)

Reply:
> That function is only called from 3 places

What about when we insert a <kbd>Ctrl+C</kbd> into the input buffer via InputStateMachineEngine?

Reply:
That goes through `HandleKeyEvent`'s off-spring `HandleGenericKeyEvent` (which was split off from it).